### DISCUSSION TITLE
How to get third party tools add themselves to the PATH?

### DISCUSSION THREAD

**Discussion by MatrixManAtYrService** [nushell/nushell#11220] [Q&A] [answered]
I use NixOS on my personal desktop, and nix-darwin on my work laptop.  In NixOS, everything works fine.  The value of `$env.PATH` is correctly set up, and I can run all of my programs.

There's [a bug](https://github.com/NixOS/nixpkgs/issues/239384) which is preventing me from using nix-darwin to install wezterm, so I've installed it via the package provided by the project maintainer.  My problem is that when I first launch the wezterm, `$PATH.env` is not adequately populated:

```console
❯ : $env.PATH
╭───┬──────────────────────────────────────────────────────╮
│ 0 │ /Users/matt/Library/Application Support/carapace/bin │
│ 1 │ /Users/matt/.bin                                     │
│ 2 │ /usr/bin                                             │
│ 3 │ /bin                                                 │
│ 4 │ /usr/sbin                                            │
│ 5 │ /sbin                                                │
╰───┴──────────────────────────────────────────────────────╯
```

I can fix this by launching zsh.
```console
❯ echo $PATH | tr ':' '\n'                                                                                        ~/src/configs
/usr/local/Caskroom/miniconda/base/condabin
/Users/matt/.docker/bin
/Users/matt/.pyenv/shims
/Users/matt/Downloads/google-cloud-sdk/bin
/Users/matt/.cargo/bin
/Users/matt/.nix-profile/bin
/etc/profiles/per-user/matt/bin
/run/current-system/sw/bin
/nix/var/nix/profiles/default/bin
/usr/local/bin
/usr/bin
/usr/sbin
```

And then by launching `nu` from there:
```
❯ nu                                                                                                              ~/src/configs

❯ : $env.PATH
╭────┬──────────────────────────────────────────────────────╮
│  0 │ /Users/matt/Library/Application Support/carapace/bin │
│  1 │ /Users/matt/.bin                                     │
│  2 │ /usr/local/Caskroom/miniconda/base/condabin          │
│  3 │ /Users/matt/.docker/bin                              │
│  4 │ /Users/matt/.pyenv/shims                             │
│  5 │ /Users/matt/Downloads/google-cloud-sdk/bin           │
│  6 │ /Users/matt/.cargo/bin                               │
│  7 │ /Users/matt/.nix-profile/bin                         │
│  8 │ /etc/profiles/per-user/matt/bin                      │
│  9 │ /run/current-system/sw/bin                           │
│ 10 │ /nix/var/nix/profiles/default/bin                    │
│ 11 │ /usr/local/bin                                       │
│ 12 │ /usr/bin                                             │
│ 13 │ /usr/sbin                                            │
│ 14 │ /bin                                                 │
│ 15 │ /sbin                                                │
│ 16 │ /Users/matt/.rvm/bin                                 │
╰────┴──────────────────────────────────────────────────────╯

```
 
 I think it's `/etc/zshenv` which contains this code:
 
 ```zsh
 # Only execute this file once per shell.
# But don't clobber the environment of interactive non-login children!
if [ -n "$__ETC_ZSHENV_SOURCED" ]; then return; fi
export __ETC_ZSHENV_SOURCED=1

# Don't execute this file when running in a pure nix-shell.
if test -n "$IN_NIX_SHELL"; then return; fi

if [ -z "$__NIX_DARWIN_SET_ENVIRONMENT_DONE" ]; then
  . /nix/store/i4zmr7an2ar64bsamsdni81sdncdf93d-set-environment
fi
 ```
 
And then `. /nix/store/i4zmr7an2ar64bsamsdni81sdncdf93d-set-environment` is what actually updates my path.  This is put in place by nix-darwin.

It's kind of annoying to have to run `zsh` and then `nu` every time I start a terminal.  I end up with a nu-zsh-nu sandwich.

I think I can hack around this by just adding the necessary paths to `env.nu`, but as evidenced by all of the other stuff cluttering up my path, this is a fairly common way to ship a tool:

1. get user the executable
2. tweak their bashrc or zshrc so that it's on the path

It seems like I can't be the first to have run across this.  Is there a general strategy for this kind of thing?  At least until it becomes common for tools to provide nushell initializers alongside their zsh initializers.

**Comment by amtoine** [comment, marked as answer]
not sure i follow, if an executable is inside a directory that is not already in the `$env.PATH`, you need to add the directory to your `env.nu` :thinking: 

we could also imagine the application ship a little script that would add something like the following to `env.nu` for the user
```nushell
$env.PATH = ($env.PATH | split row (char esep) | prepend ...)
```
 :yum:

**Reply by MatrixManAtYrService** [reply]
Sorry for the convoluted question, I became less confused over time and didn't bother rewrite it once I was less confused.

In my case it turned out to be easy to do what you suggested.  Thanks.

**Reply by amtoine** [reply]
no worries at all :wink: 
glad it works for you in the end :relieved:
